File: frontend/lib/events.ts
// SSE event envelope types — discriminated union keyed on `type`.
// Mirrors backend/src/ara/models/events.py.

import type { FinalReport, ResearchPlan, SubQueryFinding } from "@/lib/types";

export type ErrorStage = "plan" | "research" | "synthesis";

export interface PlanReadyEvent {
  type: "plan_ready";
  plan: ResearchPlan;
}

export interface ResearcherStartedEvent {
  type: "researcher_started";
  sub_query_id: string;
  question: string;
}

export interface ResearcherProgressEvent {
  type: "researcher_progress";
  sub_query_id: string;
  tool_call: string;
}

export interface ResearcherCompleteEvent {
  type: "researcher_complete";
  sub_query_id: string;
  finding: SubQueryFinding;
}

export interface SynthesisTokenEvent {
  type: "synthesis_token";
  token: string;
}

export interface ReportCompleteEvent {
  type: "report_complete";
  report: FinalReport;
}

export interface ResearchErrorEvent {
  type: "error";
  stage: ErrorStage;
  message: string;
  sub_query_id?: string | null;
}

export interface CostUpdateEvent {
  type: "cost_update";
  model: string;
  input_tokens: number;
  output_tokens: number;
  cumulative_usd: number;
}

export type ResearchEvent =
  | PlanReadyEvent
  | ResearcherStartedEvent
  | ResearcherProgressEvent
  | ResearcherCompleteEvent
  | SynthesisTokenEvent
  | ReportCompleteEvent
  | ResearchErrorEvent
  | CostUpdateEvent;

File: backend/src/ara/models/events.py
"""SSE event envelopes — a discriminated union keyed on `type`.

These are the wire format between backend and frontend. Every event published
to the per-report ring buffer is one of these models; `ResearchEventAdapter`
validates/serializes the union.
"""

from __future__ import annotations

from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, Field, TypeAdapter

from ara.models.research import FinalReport, ResearchPlan, SubQueryFinding

ErrorStage = Literal["plan", "research", "synthesis"]


class PlanReady(BaseModel):
    type: Literal["plan_ready"] = "plan_ready"
    plan: ResearchPlan


class ResearcherStarted(BaseModel):
    type: Literal["researcher_started"] = "researcher_started"
    sub_query_id: UUID
    question: str


class ResearcherProgress(BaseModel):
    type: Literal["researcher_progress"] = "researcher_progress"
    sub_query_id: UUID
    tool_call: str


class ResearcherComplete(BaseModel):
    type: Literal["researcher_complete"] = "researcher_complete"
    sub_query_id: UUID
    finding: SubQueryFinding


class SynthesisToken(BaseModel):
    type: Literal["synthesis_token"] = "synthesis_token"
    token: str


class ReportComplete(BaseModel):
    type: Literal["report_complete"] = "report_complete"
    report: FinalReport


class ErrorEvent(BaseModel):
    """An error surfaced from any stage.

    `sub_query_id` is set for researcher-scoped failures so the frontend can
    mark the right agent as degraded without killing the whole report.
    """

    type: Literal["error"] = "error"
    stage: ErrorStage
    message: str
    sub_query_id: UUID | None = None


class CostUpdate(BaseModel):
    """Running API cost snapshot; emitted after every Anthropic response."""

    type: Literal["cost_update"] = "cost_update"
    model: str
    input_tokens: int
    output_tokens: int
    cumulative_usd: float


ResearchEvent = Annotated[
    (
        PlanReady
        | ResearcherStarted
        | ResearcherProgress
        | ResearcherComplete
        | SynthesisToken
        | ReportComplete
        | ErrorEvent
        | CostUpdate
    ),
    Field(discriminator="type"),
]

ResearchEventAdapter: TypeAdapter[ResearchEvent] = TypeAdapter(ResearchEvent)
